### PR TITLE
units: Sort systemd presets in alphabetical order in 20-ignition.preset file

### PR DESCRIPTION
**Pull request by marmijo**
The 20-ignition.preset file is used in CoreOS layering. The contents of the file will be sorted in alphabetical order by unit name to ensure the file is written in a consistent order across multiple runs. 

Fixes https://github.com/coreos/ignition/issues/1339

### CODE REVIEW

File: internal/exec/stages/files/units.go [hunk 1]
```diff
@@ -151,7 +152,16 @@ func (s *stage) createSystemdPresetFile(presets map[string]*Preset) error {
 		return err
 	}
 	hasInstanceUnit := false
-	for _, value := range presets {
+
+	// sort the units before writing to the systemd presets file
```
Comment (bgilbert) [inline internal/exec/stages/files/units.go:156]: It's not completely obvious why we should do the sorting, so let's make sure to explain here.

File: internal/exec/stages/files/units.go [hunk 2]
```diff
@@ -151,7 +152,16 @@ func (s *stage) createSystemdPresetFile(presets map[string]*Preset) error {
...
+	unitNames := make([]string, 0, len(presets))
+	for unit, _ := range(presets) {
+		unitNames = append(unitNames, unit)
+	}
+	sort.Strings(unitNames)
+
+	for i := range unitNames {
```
Comment (bgilbert) [inline internal/exec/stages/files/units.go:163]: Since we only need the values, you can do:

```suggestion
	for _, name := range unitNames {
```

**Review by bgilbert — COMMENTED** [review]
For the record: let's add a positive blackbox test for this.  We'll know we've regressed if it _sometimes_ fails.

For the commit message, we usually add a context prefix.  In this case, something like "files:" is probably fine, since this change is scoped to the files stage.  Also, while it's good to link to the bug, we try to write the commit message with enough detail that the change is clear from the message alone.

File: internal/exec/stages/files/units.go (hunk 1 above)
Comment (marmijo) [reply internal/exec/stages/files/units.go]: I updated the comment to be more descriptive.

File: tests/positive/files/units.go [hunk 3]
```diff
@@ -320,3 +321,57 @@ func RemoveEnablementSymLinksforUnit() types.Test {
 		ConfigMinVersion: configMinVersion,
 	}
 }
+
+// testPresetsFileinAlphabeticalOrder checks if Ignition
+// writes the systemd presets in alphabeitcal order by
+// unit name into the 20-ignition.preset file.
+func testPresetsFileInAlphabeticalOrder() types.Test {
```
Comment (bgilbert) [inline tests/positive/files/units.go:328]: You're right that this doesn't need to be exported (initial capital letter), but apparently that's what we've been doing (?), so let's do it here too for consistency.

File: tests/positive/files/units.go [hunk 4]
```diff
@@ -320,3 +321,57 @@ func RemoveEnablementSymLinksforUnit() types.Test {
 		ConfigMinVersion: configMinVersion,
 	}
 }
+
+// testPresetsFileinAlphabeticalOrder checks if Ignition
+// writes the systemd presets in alphabeitcal order by
```
Comment (bgilbert) [inline tests/positive/files/units.go:326]: `alphabetical`

File: tests/positive/files/units.go [hunk 5]
```diff
@@ -320,3 +321,57 @@ func RemoveEnablementSymLinksforUnit() types.Test {
...
+// unit name into the 20-ignition.preset file.
+func testPresetsFileInAlphabeticalOrder() types.Test {
+	name := "unit.presets.alphabetical.order"
```
Comment (bgilbert) [inline tests/positive/files/units.go:329]: The dots form a notional hierarchy of tests, and we probably won't have another test under `unit.presets.alphabetical`.  Maybe something like `unit.presets.sorted`?

File: tests/positive/files/units.go [hunk 6]
```diff
@@ -320,3 +321,57 @@ func RemoveEnablementSymLinksforUnit() types.Test {
...
+					"enabled": true,
+					"name": "qux.service"
+				},
+				{
+					"enabled": true,
+					"name": "baz.service"
+				},
+				{
+					"enabled": true,
+					"name": "garply.service"
```
Comment (bgilbert) [inline tests/positive/files/units.go:354]: :+1:

**Review by bgilbert — COMMENTED** [review]
Looks good!

**Comment by bgilbert** [comment]
The first paragraph of the commit message is still a bit vague about why this matters.  And I think I misled you a bit wrt CoreOS layering; that's a downstream FCOS/RHCOS concept that isn't _directly_ relevant to Ignition upstream.  It probably makes more sense to refer to `ignition-apply`, which is the Ignition entrypoint used by the layering code.

I'm sorry to say I took the easy way out here, and rewrote the paragraph.  Feel free to use this verbatim or rework to taste:

```
The lines of the 20-ignition.preset file were being written in arbitrary
order.  This shouldn't affect its semantics, but could cause ignition-apply
to produce different filesystem trees from identical configs, breaking
reproducible container builds.  Sort the list of units before writing it
out.
```

File: tests/positive/files/units.go (hunk 4 above)
Comment (marmijo) [reply tests/positive/files/units.go]: fixed!

File: tests/positive/files/units.go (hunk 3 above)
Comment (marmijo) [reply tests/positive/files/units.go]: Thanks for catching this. I've updated the name.

File: tests/positive/files/units.go (hunk 5 above)
Comment (marmijo) [reply tests/positive/files/units.go]: I took a wild guess on the naming convention here. Thanks for the feedback, it's been updated.

**Review by bgilbert — APPROVED** [review]
:tada:

**Pull request merged**